### ISSUE TITLE
PyrSymbol's length is only 8bits.

### ISSUE THREAD

**Issue by JordanHendersonMusic** [supercollider/supercollider#6214]
## Steps to reproduce

```supercollider
// works, produces a file with 255 'a's
a = ('a'!255).reduce('++').asString.asSymbol
f = File("~/tmp/test.txt".standardizePath, "w")
f.write(a)
f.close

// does not work, produces an empty file
a = ('a'!256).reduce('++').asString.asSymbol
f = File("~/tmp/test.txt".standardizePath, "w")
f.write(a)
f.close
```

Here is at least one bug I have found caused by it. Most code calls `strlen` directly so doesn't suffer from this.
Luckily, the value doesn't appear to overflow, it just reverts to 0 if greater than 255.

The issue is in PyrSymbol.h

**Comment by smoge** [comment]
This issue he may be related to the length field of the PyrSymbol structure:
```


struct PyrSymbol {
    char* name;
    long hash;
    short specialIndex;
    uint8 flags;
    uint8 length;
    union {
        intptr_t index; // index in row table or primitive table
        struct PyrClass* classobj; // pointer to class with this name.
        char* source; // source code for sym_Filename; used only during compilation.
    } u;
    struct classdep* classdep;
};
```

`uint8 length`  field is defined as an unsigned 8-bit integer which can represent values from 0 to 255. If you attempt to create a symbol with a name longer than 255 characters, this field cannot  represent the length of the symbol. But that's very edgy case. 

Does it worth changing the the lenght field of symbols? I would benchmark to see check performance, and also run tests.

**Comment by JordanHendersonMusic** [comment]
I did do a PR that fixed that field... but is has a rather large knock on effect so I deleted it.

I think a more conservative approach would be to use the deprecated attribute and just change places where the length is accessed  to use `strlen`.

```c++
struct PyrSymbol {
    char* name;
    long hash;
    short specialIndex;
    uint8 flags;
    [[depreciated]] uint8 length;
    union {
        intptr_t index; // index in row table or primitive table
        struct PyrClass* classobj; // pointer to class with this name.
        char* source; // source code for sym_Filename; used only during compilation.
    } u;
    struct classdep* classdep;
};
```

**Comment by smoge** [comment]
One of the ideas that I was discussing recently was to optimize using 128 bit slots since sclang could store objects in multiple registers and use multiple instructions to do arithmetic on those values. It can work, MAYBE.

Before your issue, I never thought of such long symbols before. 

The lack of robust tests scares me a bit too.

**Comment by smoge** [comment]
> deprecated attribute and

Indicates that the use of the name or entity declared with this attribute is allowed, but discouraged for some reason. Compilers typically issue warnings on such uses. The string-literal, if specified, is usually included in the warnings.
This attribute is allowed in declarations of the following names or entities:
	•	class/struct/union, e.g., struct [[deprecated]] S;,
	•	typedef-name, including those declared by alias declaration, e.g.,
	•	[[deprecated]] typedef S* PS;,
	•	using PS [[deprecated]] = S*;,
	•	(non-member) variable, e.g., [[deprecated]] int x;,
	•	static data member, e.g., struct S { [[deprecated]] static constexpr char CR{13}; };,
	•	non-static data member, e.g., union U { [[deprecated]] int n; };,
	•	function, e.g., [[deprecated]] void f();,
	•	namespace, e.g., namespace [[deprecated]] NS { int x; },
	•	enumeration, e.g., enum [[deprecated]] E {};,
	•	enumerator, e.g., enum { A [[deprecated]], B [[deprecated]] = 42 };,
(since C++17)
	•	template specialization, e.g., template<> struct [[deprecated]] X<int> {};.
A name declared non-deprecated may be redeclared deprecated. A name declared deprecated cannot be un-deprecated by redeclaring it without this attribute.

https://en.cppreference.com/w/cpp/language/attributes/deprecated

**Comment by JordanHendersonMusic** [comment]
What I found that was even more interesting is that the file in the original post was empty, not just the wrong size. If I ask super collider for the size of the really long symbol it says it is 0. As far as I can tell this is the implementation specific result of casting a value that doesn't fit into the unit8 range, meaning what it does on other machines is anyone's guess. 

The larger issue is the lack of fixed width integers. `int` need technical be only 16 bits, yet it is assumed to be 32 throughout the code.

Converting everything to fixed width would be fun...

**Comment by JordanHendersonMusic** [comment]
Sure, but there's no validation to stop this. Also, the majority of primitives just use strlen and don't access the size member, as if this is a known issue.

**Comment by smoge** [comment]
I wonder, for example, if node messaging in granular-synthesis-like situations would be optimized using symbols rather than strings. Or in dictionary hashes

```
{ 50000000.do({ (\granularsynthesislike).hash }); }.bench; //time to run: 2.140363561 seconds.
{ 50000000.do({ ("granularsynthesislike").hash}); }.bench; // time to run: 3.133667234 seconds.
```

**Comment by telephon** [comment]
This optimisation probably comes from the fact that symbols are unique?

**Comment by smoge** [comment]
> This optimisation probably comes from the fact that symbols are unique?

Also, this. But would changing `uint8 length` to `uint32` affect performance? That would change how memory and CPU ache work, for example. But we should keep in mind that this made more sense 20 years ago, which is the context of the codebase.

Today, there are implementations where arbitrary long Integers perform more or less similarly. https://iohk.io/en/blog/posts/2020/07/28/improving-haskells-big-numbers-support/

**Comment by JordanHendersonMusic** [comment]
In many primitives and places the code stores a symbol in a temporary char[256], meaning there'd be a branch and an optional heap allocation for large symbols.

**Comment by smoge** [comment]
> In many primitives and places the code stores a symbol in a temporary char[256], meaning there'd be a branch and an optional heap allocation for large symbols.

Yes, following the principle of keeping the overhead at a minimum sounds very good, that's how those algorithms do it.

**Comment by smoge** [comment]
> This optimisation probably comes from the fact that symbols are unique?

You're right about it; this is a considerable optimization. If @JordanHendersonMusic has the code ready, let's benchmark it.

**Comment by smoge** [comment]
On second thought, you both are correct. Symbol optimization comes from allowing faster comparison and lookup, as pointers. This is the main factor. If the code already requires conditional logic to handle symbols that don't fit into an 8-bit representation, a 16-bit representation will avoid checks, and memory can become more predictable, simpler buffer allocation reduces the risk of overflow compared to variable-length encoding.

It would be worth trying? Was that the purpose of your other patch?

**Comment by telephon** [comment]
should we close?

**Comment by JordanHendersonMusic** [comment]
I can't see a solution mentioned here? 

I can see two options:

1.  Symbols should be limited to 256 chars and class names to 250 (256 - "Meta_"), and this should be enforced, meaning all code than currently calls strlen can now access the length field.

2. Symbols should allow sizes up to size_t. This means a big change to many many parts of the code that only consider the first 256 characters.

I'm in favour of the first, but there are places where long symbols might occur (OSC addresses could get long?).

**Comment by smoge** [comment]
> I'm in favour of the first, but there are places where long symbols might occur (OSC addresses could get long?).

For example, when I wrote the ArdourOSC interface, I found nothing that long. With OSC, one can control _any_ aspect of this DAW. I don't know if there are such practical cases.

**Comment by smoge** [comment]
Can you find a practical case? I've never seen it this long.

`/This/Is/An/Example/Of/A/Very/Long/OSC/Message/That/Is/Designed/To/Just/Slightly/Exceed/The/Limit/Of/Two/Hundred/And/Sixty/Characters/To/Show/How/SuperCollider/Could/Accommodate/Such/Lengths/Without/Compromising/On/Implementation/Or/Integrity/In/OSC/Messages`

**Comment by dannyZyg** [comment]
I'm having an issue and I'm not 100% sure if it relates but it sounds like it does:

My MacOS system language is set to Japanese and what comes back from `ServerOptions.devices` does not match what is listed in the system audio devices.

e.g. `MacBook Airのスピーカー` (system) -> `MacBook Airのス` (sclang)

This prevents me from selecting some devices (at least by name).

From the above discussion I can't quite tell if there is a workaround for this particular case. Wherever `ServerOptions` gets the list, can `strlen` be used instead of `length`?

**Comment by JordanHendersonMusic** [comment]
Hi @dannyZyg , it won't be the size of the symbol that's the issue here. Would you mind opening a new issue with your problem please?

**Comment by dannyZyg** [comment]
Ah ok cool - I wanted to check before I opened an issue in case it was the same thing. Thanks for the info @JordanHendersonMusic !
